File: tools/db-seeder.ts
/**
 * Script de génération de données de test pour PartyLogger
 * Génère des étudiants, invités, soirées, tickets et logs en masse
 * 
 * Usage: npx tsx tools/db-seeder.ts
 */

import { PrismaClient } from '@prisma/client';
import { faker } from '@faker-js/faker/locale/fr';

const prisma = new PrismaClient();

// Configuration
const CONFIG = {
	STUDENTS_COUNT: 200,
	GUESTS_COUNT: 50,
	PARTIES_COUNT: 15,
};

// Données de référence
const SPECIALITES = ['ASE1', 'ASE2', 'ASE3', 'IR1', 'IR2', 'IR3', 'T&F1', 'T&F2', 'T&F3', 'Méca1', 'Méca2', 'Méca3', 'GI1', 'GI2', 'GI3'];
const PARTY_NAMES = [
	'Soirée de Rentrée',
	'Halloween Party',
	'Soirée Karaoké',
	'Soirée Jeux',
	'Nuit Blanche',
	'Soirée Déguisée',
	'After Exam Party',
	'Soirée BBQ',
	'Soirée 80s',
	'Galette des Rois',
	'Saint-Valentin',
	'Soirée Casino',
	'Beach Party',
	'Oktoberfest',
	'Nouvel An'
];
const LOCATIONS = ['BDE ENSISA', 'Foyer', 'Salle polyvalente', 'Campus extérieur', 'Partenaire bar'];

/**
 * Supprime toutes les données existantes
 */
async function clearAllData() {
	console.log('⚠️  Suppression de toutes les données existantes...');
	
	await prisma.log.deleteMany({});
	console.log('   ✓ Logs supprimés');
	
	await prisma.ticket.deleteMany({});
	console.log('   ✓ Tickets supprimés');
	
	await prisma.guest.deleteMany({});
	console.log('   ✓ Invités supprimés');
	
	await prisma.student.deleteMany({});
	console.log('   ✓ Étudiants supprimés');
	
	await prisma.partyStats.deleteMany({});
	console.log('   ✓ Stats de soirées supprimées');
	
	await prisma.party.deleteMany({});
	console.log('   ✓ Soirées supprimées');
	
	console.log('✓ Toutes les données ont été supprimées\n');
}

/**
 * Génère un numéro étudiant unique
 */
function generateUniqueStudentId(existingIds: Set<number>): number {
	let studentId: number;
	do {
		studentId = faker.number.int({ min: 20000000, max: 23999999 });
	} while (existingIds.has(studentId));
	existingIds.add(studentId);
	return studentId;
}

/**
 * Génère des étudiants
 */
async function generateStudents(count: number) {
	console.log(`Génération de ${count} étudiants...`);
	
	const existingIds = new Set<number>();
	const students = [];
	
	for (let i = 0; i < count; i++) {
		students.push({
			studentId: generateUniqueStudentId(existingIds),
			first_name: faker.person.firstName(),
			last_name: faker.person.lastName(),
			speciality: faker.helpers.arrayElement(SPECIALITES),
			isMember: faker.datatype.boolean(),
		});
	}
	
	await prisma.student.createMany({ data: students });
	
	const createdStudents = await prisma.student.findMany({ select: { id: true } });
	console.log(`✓ ${count} étudiants créés\n`);
	
	return createdStudents.map(s => s.id);
}

/**
 * Génère des soirées
 */
async function generateParties(creatorId: number, count: number) {
	console.log(`Génération de ${count} soirées...`);
	
	const baseDate = new Date();
	baseDate.setFullYear(baseDate.getFullYear() - 1); // Il y a un an
	
	const parties = [];
	
	for (let i = 0; i < count; i++) {
		let name = faker.helpers.arrayElement(PARTY_NAMES);
		if (i > 0) {
			name = `${name} ${faker.date.month()}`;
		}
		
		const description = `Une soirée ${faker.word.adjective()} organisée par le BDE`;
		const date = new Date(baseDate.getTime() + faker.number.int({ min: 0, max: 365 * 24 * 60 * 60 * 1000 }));
		const location = faker.helpers.arrayElement(LOCATIONS);
		
		const now = new Date();
		const isClosed = date < new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
		const isActive = !isClosed && date < now && Math.random() < 0.1;
		
		const startTime = new Date(date);
		startTime.setHours(20, 0, 0, 0);
		
		const endTime = isClosed ? new Date(startTime.getTime() + faker.number.int({ min: 4, max: 8 }) * 60 * 60 * 1000) : null;
		
		parties.push({
			name,
			description,
			date,
			location,
			createdBy: creatorId,
			isActive,
			isClosed,
			startTime,
			endTime,
		});
	}
	
	// Créer les soirées et leurs stats
	const createdParties = [];
	for (const party of parties) {
		const created = await prisma.party.create({
			data: {
				...party,
				stats: {
					create: {}
				}
			}
		});
		createdParties.push({
			id: created.id,
			date: created.date,
			startTime: created.startTime!,
			isClosed: created.isClosed,
			isActive: created.isActive,
		});
	}
	
	console.log(`✓ ${count} soirées créées\n`);
	return createdParties;
}

/**
 * Génère des invités
 */
async function generateGuests(studentIds: number[], count: number) {
	console.log(`Génération de ${count} invités...`);
	
	const guests = [];
	
	for (let i = 0; i < count; i++) {
		guests.push({
			first_name: faker.person.firstName(),
			last_name: faker.person.lastName(),
			guarantorId: faker.helpers.arrayElement(studentIds),
		});
	}
	
	await prisma.guest.createMany({ data: guests });
	
	const createdGuests = await prisma.guest.findMany({ select: { id: true } });
	console.log(`✓ ${count} invités créés\n`);
	
	return createdGuests.map(g => g.id);
}

/**
 * Génère des tickets et logs pour les soirées
 */
async function generateTicketsAndLogs(
	parties: Array<{ id: number; date: Date; startTime: Date; isClosed: boolean; isActive: boolean }>,
	studentIds: number[],
	guestIds: number[]
) {
	console.log('Génération des tickets et logs...');
	
	let totalTickets = 0;
	let totalLogs = 0;
	
	for (const party of parties) {
		const partyId = party.id;
		const partyDate = party.date;
		const isClosed = party.isClosed;
		
		// Nombre de participants (30% à 80% des étudiants)
		const numStudents = faker.number.int({ 
			min: Math.floor(studentIds.length * 0.3), 
			max: Math.floor(studentIds.length * 0.8) 
		});
		const participatingStudents = faker.helpers.arrayElements(studentIds, numStudents);
		
		// Nombre d'invités (5% à 15%)
		const numGuests = faker.number.int({ 
			min: Math.floor(guestIds.length * 0.05), 
			max: Math.floor(guestIds.length * 0.15) 
		});
		const participatingGuests = faker.helpers.arrayElements(guestIds, Math.min(numGuests, guestIds.length));
		
		// Tickets étudiants
		for (const studentId of participatingStudents) {
			const entryTime = new Date(
				partyDate.getTime() + 
				faker.number.int({ min: 0, max: 3 * 60 * 60 * 1000 }) + 
				faker.number.int({ min: 0, max: 59 * 60 * 1000 })
			);
			
			// 70% restent jusqu'à la fin
			let exitTime = entryTime;
			if (isClosed && Math.random() < 0.7) {
				exitTime = new Date(entryTime.getTime() + faker.number.int({ min: 2, max: 6 }) * 60 * 60 * 1000);
			}
			
			await prisma.ticket.create({
				data: {
					student: { connect: { id: studentId } },
					party: { connect: { id: partyId } },
					entryAt: entryTime,
					exitAt: exitTime,
					createdAt: entryTime,
				}
			});
			totalTickets++;
			
			// Log d'entrée
			await prisma.log.create({
				data: {
					student: { connect: { id: studentId } },
					party: { connect: { id: partyId } },
					action: 'ENTRY',
					timestamp: entryTime,
				}
			});
			totalLogs++;
			
			// Log de sortie si parti
			if (exitTime.getTime() !== entryTime.getTime()) {
				await prisma.log.create({
					data: {
						student: { connect: { id: studentId } },
						party: { connect: { id: partyId } },
						action: 'EXIT',
						timestamp: exitTime,
					}
				});
				totalLogs++;
			}
		}
		
		// Tickets invités
		for (const guestId of participatingGuests) {
			const entryTime = new Date(
				partyDate.getTime() + 
				faker.number.int({ min: 0, max: 3 * 60 * 60 * 1000 }) + 
				faker.number.int({ min: 0, max: 59 * 60 * 1000 })
			);
			
			let exitTime = entryTime;
			if (isClosed && Math.random() < 0.7) {
				exitTime = new Date(entryTime.getTime() + faker.number.int({ min: 2, max: 6 }) * 60 * 60 * 1000);
			}
			
			await prisma.ticket.create({
				data: {
					guest: { connect: { id: guestId } },
					party: { connect: { id: partyId } },
					entryAt: entryTime,
					exitAt: exitTime,
					createdAt: entryTime,
				}
			});
			totalTickets++;
			
			// Log d'entrée
			await prisma.log.create({
				data: {
					guest: { connect: { id: guestId } },
					party: { connect: { id: partyId } },
					action: 'ENTRY',
					timestamp: entryTime,
				}
			});
			totalLogs++;
			
			// Log de sortie si parti
			if (exitTime.getTime() !== entryTime.getTime()) {
				await prisma.log.create({
					data: {
						guest: { connect: { id: guestId } },
						party: { connect: { id: partyId } },
						action: 'EXIT',
						timestamp: exitTime,
					}
				});
				totalLogs++;
			}
		}
		
		// Mettre à jour les stats de la soirée
		await prisma.partyStats.update({
			where: { partyId },
			data: {
				totalTickets: numStudents + numGuests,
				totalStudents: numStudents,
				totalGuests: numGuests,
			}
		});
	}
	
	console.log(`✓ ${totalTickets} tickets créés`);
	console.log(`✓ ${totalLogs} logs créés\n`);
}

/**
 * Point d'entrée principal
 */
async function main() {
	console.log('='.repeat(60));
	console.log('   GÉNÉRATEUR DE DONNÉES DE TEST - PartyLogger');
	console.log('='.repeat(60));
	console.log();
	
	console.log('⚠️  ATTENTION: Ce script va supprimer TOUTES les données existantes');
	console.log('   et les remplacer par des données de test générées aléatoirement.');
	console.log();
	console.log('Configuration:');
	console.log(`  • ${CONFIG.STUDENTS_COUNT} étudiants`);
	console.log(`  • ${CONFIG.GUESTS_COUNT} invités`);
	console.log(`  • ${CONFIG.PARTIES_COUNT} soirées`);
	console.log();
	
	// En TypeScript, on ne peut pas facilement demander confirmation en CLI
	// Donc on supprime directement (décommenter la ligne suivante pour activer)
	console.log('⏸️  Pour éviter la suppression accidentelle, commentez clearAllData() dans le code');
	console.log();
	
	try {
		console.log('✓ Connexion à la base de données établie\n');
		
		// Supprimer les données existantes
		await clearAllData();
		
		// Obtenir l'ID du premier utilisateur (admin)
		const firstUser = await prisma.user.findFirst({ select: { id: true } });
		if (!firstUser) {
			console.log('❌ Aucun utilisateur trouvé. Créez d\'abord un utilisateur admin.');
			return;
		}
		const creatorId = firstUser.id;
		
		// Générer les données
		console.log('Démarrage de la génération de données...\n');
		
		const studentIds = await generateStudents(CONFIG.STUDENTS_COUNT);
		const guestIds = await generateGuests(studentIds, CONFIG.GUESTS_COUNT);
		const parties = await generateParties(creatorId, CONFIG.PARTIES_COUNT);
		await generateTicketsAndLogs(parties, studentIds, guestIds);
		
		console.log('='.repeat(60));
		console.log('✅ GÉNÉRATION TERMINÉE AVEC SUCCÈS!');
		console.log('='.repeat(60));
		console.log('\nRésumé:');
		console.log(`  • ${studentIds.length} étudiants`);
		console.log(`  • ${guestIds.length} invités`);
		console.log(`  • ${parties.length} soirées`);
		console.log('  • Tickets et logs générés automatiquement');
		console.log();
		
	} catch (error) {
		console.error('\n❌ Erreur:', error);
		throw error;
	} finally {
		await prisma.$disconnect();
	}
}

main()
	.catch((error) => {
		console.error(error);
		process.exit(1);
	});

File: tools/DB_Seeder.py
#!/usr/bin/env python3
"""
Script de génération de données de test pour PartyLogger
Génère des étudiants, invités, soirées, tickets et logs en masse
"""

import mysql.connector
from datetime import datetime, timedelta
import random
from faker import Faker

# Configuration de la base de données
DB_CONFIG = {
    'host': '45.147.97.78',
    'port': 3306,
    'user': 'admin',
    'password': 'admin123',
    'database': 'students'
}

# Initialisation de Faker pour générer des données réalistes
fake = Faker('fr_FR')

# Listes pour les données
SPECIALITES = ['ASE1', 'ASE2', 'ASE3', 'IR1', 'IR2', 'IR3', 'T&F1', 'T&F2', 'T&F3', 'Méca1', 'Méca2', 'Méca3', 'GI1', 'GI2', 'GI3']
PARTY_NAMES = [
    'Soirée de Rentrée',
    'Halloween Party',
    'Soirée Karaoké',
    'Soirée Jeux',
    'Nuit Blanche',
    'Soirée Déguisée',
    'After Exam Party',
    'Soirée BBQ',
    'Soirée 80s',
    'Galette des Rois',
    'Saint-Valentin',
    'Soirée Casino',
    'Beach Party',
    'Oktoberfest',
    'Nouvel An'
]
LOCATIONS = ['BDE ENSISA', 'Foyer', 'Salle polyvalente', 'Campus extérieur', 'Partenaire bar']

def get_connection():
    """Établit une connexion à la base de données"""
    return mysql.connector.connect(**DB_CONFIG)

def clear_all_data(conn):
    """Supprime toutes les données existantes (ATTENTION: opération destructive)"""
    cursor = conn.cursor()
    print("⚠️  Suppression de toutes les données existantes...")
    
    # Désactiver les contraintes de clés étrangères temporairement
    cursor.execute("SET FOREIGN_KEY_CHECKS = 0")
    
    # Supprimer dans l'ordre inverse des dépendances
    tables = ['Log', 'Ticket', 'Guest', 'Student', 'PartyStats', 'Party']
    for table in tables:
        cursor.execute(f"DELETE FROM {table}")
        print(f"   ✓ {table} vidé")
    
    # Réactiver les contraintes
    cursor.execute("SET FOREIGN_KEY_CHECKS = 1")
    conn.commit()
    print("✓ Toutes les données ont été supprimées\n")

def generate_students(conn, count=200):
    """Génère des étudiants"""
    cursor = conn.cursor()
    print(f"Génération de {count} étudiants...")
    
    students = []
    student_ids = set()
    
    for _ in range(count):
        # Générer un numéro étudiant unique
        while True:
            student_id = random.randint(20000000, 23999999)
            if student_id not in student_ids:
                student_ids.add(student_id)
                break
        
        first_name = fake.first_name()
        last_name = fake.last_name()
        speciality = random.choice(SPECIALITES)
        is_member = random.choice([True, False])
        
        cursor.execute(
            "INSERT INTO Student (studentId, first_name, last_name, speciality, isMember) VALUES (%s, %s, %s, %s, %s)",
            (student_id, first_name, last_name, speciality, is_member)
        )
        students.append(cursor.lastrowid)
    
    conn.commit()
    print(f"✓ {count} étudiants créés\n")
    return students

def generate_parties(conn, creator_id, count=15):
    """Génère des soirées"""
    cursor = conn.cursor()
    print(f"Génération de {count} soirées...")
    
    parties = []
    base_date = datetime.now() - timedelta(days=365)  # Commence il y a un an
    
    for i in range(count):
        name = random.choice(PARTY_NAMES)
        if i > 0:
            name = f"{name} {fake.month_name()}"
        
        description = f"Une soirée {fake.word()} organisée par le BDE"
        date = base_date + timedelta(days=random.randint(0, 365))
        location = random.choice(LOCATIONS)
        
        # Déterminer le statut
        is_closed = date < datetime.now() - timedelta(days=7)
        is_active = not is_closed and date < datetime.now() and random.random() < 0.1
        
        start_time = date.replace(hour=20, minute=0, second=0)
        end_time = start_time + timedelta(hours=random.randint(4, 8)) if is_closed else None
        
        cursor.execute(
            """INSERT INTO Party (name, description, date, location, creatorId, isActive, isClosed, startTime, endTime) 
               VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)""",
            (name, description, date, location, creator_id, is_active, is_closed, start_time, end_time)
        )
        party_id = cursor.lastrowid
        
        # Créer les stats pour cette soirée
        cursor.execute(
            "INSERT INTO PartyStats (partyId) VALUES (%s)",
            (party_id,)
        )
        
        parties.append({
            'id': party_id,
            'date': date,
            'start_time': start_time,
            'is_closed': is_closed,
            'is_active': is_active
        })
    
    conn.commit()
    print(f"✓ {count} soirées créées\n")
    return parties

def generate_guests(conn, student_ids, count=50):
    """Génère des invités"""
    cursor = conn.cursor()
    print(f"Génération de {count} invités...")
    
    guests = []
    
    for _ in range(count):
        first_name = fake.first_name()
        last_name = fake.last_name()
        guarantor_id = random.choice(student_ids)
        
        cursor.execute(
            "INSERT INTO Guest (first_name, last_name, guarantorId) VALUES (%s, %s, %s)",
            (first_name, last_name, guarantor_id)
        )
        guests.append(cursor.lastrowid)
    
    conn.commit()
    print(f"✓ {count} invités créés\n")
    return guests

def generate_tickets_and_logs(conn, parties, student_ids, guest_ids):
    """Génère des tickets et logs pour les soirées"""
    cursor = conn.cursor()
    print("Génération des tickets et logs...")
    
    total_tickets = 0
    total_logs = 0
    
    for party in parties:
        party_id = party['id']
        party_date = party['date']
        is_closed = party['is_closed']
        
        # Nombre de participants pour cette soirée (30% à 80% des étudiants)
        num_students = random.randint(int(len(student_ids) * 0.3), int(len(student_ids) * 0.8))
        participating_students = random.sample(student_ids, num_students)
        
        # Nombre d'invités (5% à 15% des invités)
        num_guests = random.randint(int(len(guest_ids) * 0.05), int(len(guest_ids) * 0.15))
        participating_guests = random.sample(guest_ids, min(num_guests, len(guest_ids)))
        
        # Tickets étudiants
        for student_id in participating_students:
            entry_time = party_date + timedelta(
                hours=random.randint(0, 3),
                minutes=random.randint(0, 59)
            )
            
            # 70% des gens sont restés jusqu'à la fin
            if is_closed and random.random() < 0.7:
                exit_time = entry_time + timedelta(hours=random.randint(2, 6))
            else:
                exit_time = entry_time  # Encore présent ou non parti
            
            cursor.execute(
                """INSERT INTO Ticket (studentId, partyId, entryAt, exitAt, createdAt, updatedAt) 
                   VALUES (%s, %s, %s, %s, %s, %s)""",
                (student_id, party_id, entry_time, exit_time, entry_time, exit_time)
            )
            total_tickets += 1
            
            # Log d'entrée
            cursor.execute(
                """INSERT INTO Log (studentId, partyId, action, timestamp) 
                   VALUES (%s, %s, %s, %s)""",
                (student_id, party_id, 'ENTRY', entry_time)
            )
            total_logs += 1
            
            # Log de sortie si parti
            if exit_time != entry_time:
                cursor.execute(
                    """INSERT INTO Log (studentId, partyId, action, timestamp) 
                       VALUES (%s, %s, %s, %s)""",
                    (student_id, party_id, 'EXIT', exit_time)
                )
                total_logs += 1
        
        # Tickets invités
        for guest_id in participating_guests:
            entry_time = party_date + timedelta(
                hours=random.randint(0, 3),
                minutes=random.randint(0, 59)
            )
            
            if is_closed and random.random() < 0.7:
                exit_time = entry_time + timedelta(hours=random.randint(2, 6))
            else:
                exit_time = entry_time
            
            cursor.execute(
                """INSERT INTO Ticket (guestId, partyId, entryAt, exitAt, createdAt, updatedAt) 
                   VALUES (%s, %s, %s, %s, %s, %s)""",
                (guest_id, party_id, entry_time, exit_time, entry_time, exit_time)
            )
            total_tickets += 1
            
            # Log d'entrée
            cursor.execute(
                """INSERT INTO Log (guestId, partyId, action, timestamp) 
                   VALUES (%s, %s, %s, %s)""",
                (guest_id, party_id, 'ENTRY', entry_time)
            )
            total_logs += 1
            
            # Log de sortie si parti
            if exit_time != entry_time:
                cursor.execute(
                    """INSERT INTO Log (guestId, partyId, action, timestamp) 
                       VALUES (%s, %s, %s, %s)""",
                    (guest_id, party_id, 'EXIT', exit_time)
                )
                total_logs += 1
        
        # Mettre à jour les stats de la soirée
        cursor.execute(
            """UPDATE PartyStats 
               SET totalTickets = %s, totalStudents = %s, totalGuests = %s 
               WHERE partyId = %s""",
            (num_students + num_guests, num_students, num_guests, party_id)
        )
    
    conn.commit()
    print(f"✓ {total_tickets} tickets créés")
    print(f"✓ {total_logs} logs créés\n")

def main():
    """Point d'entrée principal"""
    print("=" * 60)
    print("   GÉNÉRATEUR DE DONNÉES DE TEST - PartyLogger")
    print("=" * 60)
    print()
    
    # Demander confirmation
    print("⚠️  ATTENTION: Ce script va supprimer TOUTES les données existantes")
    print("   et les remplacer par des données de test générées aléatoirement.")
    print()
    response = input("Voulez-vous continuer? (oui/non): ")
    
    if response.lower() not in ['oui', 'o', 'yes', 'y']:
        print("\n❌ Opération annulée")
        return
    
    print("\n" + "=" * 60)
    print()
    
    try:
        conn = get_connection()
        print("✓ Connexion à la base de données établie\n")
        
        # Supprimer les données existantes
        clear_all_data(conn)
        
        # Obtenir l'ID du premier utilisateur (admin) pour créer les soirées
        cursor = conn.cursor()
        cursor.execute("SELECT id FROM User LIMIT 1")
        result = cursor.fetchone()
        if not result:
            print("❌ Aucun utilisateur trouvé. Créez d'abord un utilisateur admin.")
            return
        creator_id = result[0]
        
        # Générer les données
        print("Démarrage de la génération de données...\n")
        
        student_ids = generate_students(conn, count=200)
        guest_ids = generate_guests(conn, student_ids, count=50)
        parties = generate_parties(conn, creator_id, count=15)
        generate_tickets_and_logs(conn, parties, student_ids, guest_ids)
        
        print("=" * 60)
        print("✅ GÉNÉRATION TERMINÉE AVEC SUCCÈS!")
        print("=" * 60)
        print(f"\nRésumé:")
        print(f"  • {len(student_ids)} étudiants")
        print(f"  • {len(guest_ids)} invités")
        print(f"  • {len(parties)} soirées")
        print(f"  • Tickets et logs générés automatiquement")
        print()
        
        conn.close()
        
    except mysql.connector.Error as err:
        print(f"\n❌ Erreur MySQL: {err}")
    except Exception as e:
        print(f"\n❌ Erreur: {e}")

if __name__ == "__main__":
    main()
